tools/launcher/cmonitor_launcher.py

```diff
@@ -0,0 +1,237 @@
+            for value in tmp_ip:
+                self.process_host_dict[key] = value
+                tmp_ip.remove(value)
+        # Printing resultant dictionary
+        print("Input [filter-host]: " + str(self.process_host_dict))
+        logging.info(f"Input [filter-host] {str(self.process_host_dict)}")
+
+    def process_events(self, event):
+        """Main thread function for processing input events from the queue.
+        Args:
+          event: events to read from this queue.
+          The events from this queue will be processed by this threading function to
+          launch cMonitor with appropriate command input.
+
+        """
+        try:
+            entry = 1
+            while True:
+                if not event.empty():
+                    fileList = event.get()
+                    for key, value in fileList.items():
+                        filename = key
+                        process_name = value
+                    logging.info(f"In processing event from the Queue - event: {entry},file: {filename},process_name: {process_name}")
+                    logging.info(f"Launching cMonitor with: {filename} with IP :{self.process_host_dict.get(process_name)}")
+                    # self.__launch_cmonitor(filename, self.process_host_dict.get(process_name))
+                    self.__launch_cmonitor(filename, process_name)
+                    entry = entry + 1
+                else:
+                    time.sleep(self.timeout)
+                    logging.info(f"In processing event Queue is empty - sleeping: {self.timeout} sec")
+                    if cmonitor_watcher.exit_flag is True:
+                        logging.info(f"In processing event_flag set to {cmonitor_watcher.exit_flag}")
+                        exit(1)
+
+        except event.Empty():
+            pass
+
+    def __launch_cmonitor(self, filename, process_name):
+        """
+        - Lauch cMonitor with appropriate command.
+        """
+
+        for c in self.command:
+            cmd = c.strip()
+            ip = self.process_host_dict.get(process_name)
+            ip_port = ip.split(":")
+            ip = ip_port[0]
+            port = ip_port[1]
+            base_path = os.path.dirname(filename)
+            path = "/".join(base_path.split("/")[5:])
+            monitor_cmd = f"{cmd} --cgroup-name={path} --remote-ip {ip} --remote-port {port}"
+            monitor_cmd = [
+                f"{cmd}",
+                f"--cgroup-name={path}",
+                f"--remote-ip {ip}",
+                f"--remote-port {port}",
+            ]
+            print("Launch cMonitor with command:", monitor_cmd)
+            logging.info(f"Launch cMonitor with command: {monitor_cmd }")
+            # monitor_process = subprocess.Popen(monitor_cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            monitor_process = subprocess.Popen(monitor_cmd, shell=True)
+            self.monitored_processes[process_name] = monitor_process
```

Comment:
my understanding of this class CmonitorLauncher is that it requires the caller to provide a static assigment "process_name -> IP address/port used by cmonitor with Prometheus frontend" right?
What if the same process_name is detected in e.g. 10 container instances? I think here we will overwrite the self.monitored_processes[process_name] each time with a new subprocess.Popen() instance, leaking the previous instance.
I think we need a more flexible IP address/port allocation scheme (when launching a cmonitor_collector configured to stream data to Prometheus) or more flexible JSON output name allocation scheme (when launching a cmonitor_collector configured to save data to JSON -- probably not a concern in this first implementation).
Perhaps we can have a meeting to discuss this but I suggest to ask to the user (through CLI options) a range of IPs/ports that can be used. Then CMonitorLauncher should have a pooling mechanism that keeps track of free IPadress/port pairs and used ones. When a new cmonitor_collector is launched we ask for next free IPaddress/port. When a cmonitor_launcher completes, we must release his IPaddress/port back to the pool.